vtex/ab-tester
# Add verification to recent data gathered before use it to update ABTestParameters

#### What is the purpose of this pull request?
Detect awkward situations for two compared workspaces where theirs conversion rate don't behave like it's expected and then act restarting the A/B Test Parameters aiming to redistribute the traffic in the workspaces to check if the recent data observed can predict the actual conversion rate of some workspaces. That last part is important as we don't want to Restart the parameters if the verification returns `false` for any workspace (which means that for this workspace the data is ok) and we should keep the test flow normally. If at least one workspace has the test result correct for it, at least this workspace will have the answer if it's better or worse than `master`.

#### What problem is this solving?
Events like seasonality of visitors can distort metrics in situations where a test is being executed and the traffic of workspaces has already been modified. E.g. a distribution of 60:40 traffic can give the workspace with more traffic a benefit when it comes a period where better visitors arrive as they're distributed 60:40 between them.
This PR should allow the traffic control part of the code detect when something like that is happening and redirect traffic to verify if the change in a metric should be treated as an anomaly due to specific events that don't corresponds to the general behavior of a visitor.

#### How should this be manually tested?
It's kind of difficult to actually test this thing in production-like environment (it depends on such a awkward situation that the verification for a Restart of workspace's traffic parameters returns `true`) but it can be tested for sanity in isolated environment.

Only took the code relative to the function `RandomRestart` and then:
 - Generate fake data of workspaces with types `WorkspaceData` and `WorkspaceCompleteData` such that conversion rate differs a lot from `SinceBeginning` to `Last24Hours` - also the relative variation which can be achieved with relative variations with opposite signal;
 - Call `RandomRestart` for these `WorkspaceCompleteData` generated some times and see if the function returns true.

#### Types of changes

P.S.: This PR contains tons of refactoring and code cleaning including changes in names of files and variables. The main focus here is the file `node/workspace/modify.ts` and the function `RandomRestart`.
